packet03-redirecting/README.org

```diff
@@ -5,49 +5,277 @@
+|    eth0 (MAC=X2)   |        |    eth1 (MAC=Y2)   |
+----------||----------        ----------||----------
+    veth0 (MAC=X1)  <-----------  veth1 (MAC=Y1)
+#+end_src
+Setup the two environments, patch the =xdp_redirect= program accordingly, and
+attach it to the =right= interface.  To test load the program, enter the right
+environment, and ping the inner interface of the left environment (your IPv6
+address will be different):
+#+begin_src sh
+$ t enter -n right
+$ ping fc00:dead:cafe:10::2
+#+end_src
+Run the =tcpdump= program inside the =left= environment. You should see that
+the ping requests are delivered and ping replies are sent back. However, unless
+forwarding is enabled on the host, they won't be delivered.  (We will fix this
+in the next assignment.)
+#+begin_src sh
+$ t enter -n left
+# tcpdump -l
+listening on veth0, link-type EN10MB (Ethernet), capture size 262144 bytes
+17:03:11.455320 IP6 fc00:dead:cafe:11::2 > fc00:dead:cafe:10::2: ICMP6, echo request, seq 1, length 64
+17:03:11.455343 IP6 fc00:dead:cafe:10::2 > fc00:dead:cafe:11::2: ICMP6, echo reply, seq 1, length 64
+#+end_src
+Finally, just in case, check that the =right= environment actually redirects
+packets (=XDP_REDIRECT= row should be non-zero):
+#+begin_src sh
+$ sudo ./xdp_stats -d right
+
+Collecting stats from BPF map
+ - BPF map (bpf_map_type:6) id:183 name:xdp_stats_map key_size:4
+   value_size:16 max_entries:5
+XDP-action
+XDP_ABORTED            0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250143
+XDP_DROP               0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250180
+XDP_PASS               0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250180
+XDP_TX                 0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250179
+XDP_REDIRECT         176 pkts (         4 pps)          20 Kbytes (     0 Mbits/s) period:0.250179
+#+end_src
 
 ** Assignment 3: Extend to a bidirectional router
-For this you'll need to install redirect packets on both interfaces and make
-packets be redirected both ways. See if you can get a ping between the two
-inner interfaces to work.
+
+#Patching MAC addresses is actually the biggest challenge of this assignment.
+#To pass MAC addresses to the XDP program a new =iface_params= map is created.
+#You need to implement the =write_iface_params= function in [[file:xdp_loader_redirect.c]]
+#which populates the map.  On the receive side, implement the
+#=read_iface_params= function inside the [[file:xdp_prog_kern.c]] file.
+
+#To test the program, configure two test environments and attach the program to
+#the second interface using a new =xdp_loader_redirect= loader. This loader accepts
+#the =-r= parameter which is used to specify an interface to redirect packets to:
+##+begin_src sh
+#$ t setup -n left
+#$ t setup -n right
+#$ sudo ./xdp_loader_redirect --progsec xdp_redirect --skb -F -d right -r left
+##+end_src
```

Comment:
Looks like the commented-out bit is just left over from earlier and should be removed?

Reply:
Yes, thanks!